### PR TITLE
Add maxLength/maxItems/maxProperties CEL extraction support/automatic maxLength calculation

### PR DESCRIPTION
**Pull request by DangerOnTheRanger**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Part of https://github.com/kubernetes/kubernetes/issues/107573, particularly laying the foundation for [resource constraints](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#estimated-cost-limits) by extracting `maxLength` (or its equivalent for non-string datatypes) when given by the user, and estimating it otherwise. This PR intentionally introduces no observable change in behavior; a follow-up PR that uses this PR in conjunction with CEL's [CostEstimator](https://github.com/google/cel-go/blob/a29469fda231aa37802d476ef818e2224c25e157/checker/cost.go#L25-L37) will be written.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas_test.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas_test.go:275]: `schemaWithMaxLengths`?

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas_test.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas_test.go:242]: `TestMaxLengths`?

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:67]: Since all arrays, maps and strings will end up being assigned a limit. Do you need an extracted constant?

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:123]: What about plain strings? Shouldn't `return StringType` on line 138 return a StringType with a maxLength if it's provided?

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas_test.go
Comment (DangerOnTheRanger) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas_test.go]: Makes much more sense that way, fixed

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas_test.go
Comment (DangerOnTheRanger) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas_test.go]: Fixed

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (DangerOnTheRanger) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:67]: That's true, when every CEL-related PR merges a constant wouldn't make much sense. Maybe it'd be fine to leave it as-is because of that.

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (DangerOnTheRanger) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:123]: Gotcha, fixed

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:79]: replace with a single `if maxItems == nil { maxItems = estimateMaxSizeJSON(s) }`

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:98]: same

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:242]: This should support string as well (with bytes being a format)

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas_test.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas_test.go:242]: How about having a single EstimateMaxLengths test function that covers both the cases where a maximum is defined on the schema and the cases where it is calculated (just label in the test cases which  is which)?

**Review by jpbetz — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (DangerOnTheRanger) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go]: Also fixed

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas_test.go
Comment (DangerOnTheRanger) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas_test.go]: Definitely makes more sense that way, fixed

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (DangerOnTheRanger) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:251]: (partially fixed, will resolve conversation when test cases that cover `duration`/`date-time` are added)

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas_test.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas_test.go:264]: It's also okay to just put the calculation as the literal, e.g. `ExpectedMaxLength: (3000000 - 2) / 2)` (but keep the JSON form in the comment)

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:26]: comment on this for sure. Also reference where in the code the 3MB limit is defined if possible (and if we can reference the const value here instead of re-declaring it, that would be amazing)

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:29]: Ooh, that's better than I expected. I was thinking the json format would be like "1h30m5s"

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:251]: Thinking about this more, I'm not 100% certain the max for anything but string/bytes will end up being needed. But hopefully that becomes clear when you test.

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/types.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/types.go:34]: Is it worth using nil to represent the absence of a maxLength and just using *int64s everywhere?  I'm worried a maxLength field is going to get zero valued somewhere on accident and we're going to have a `0` where we wanted a `-1` and it's going to mess things up.  If we instead use nil, then the zero value does the right thing.

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas_test.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas_test.go:325]: :+1:

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:226]: `case "number", "integer":` (since they're the same)

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (DangerOnTheRanger) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:28]: Added a comment - I'm marking this as resolved since the TODO ought to be fixed before merging.

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (DangerOnTheRanger) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go]: Fixed!

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (DangerOnTheRanger) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go]: Duration/date/time all share the same format in OpenAPI, so I think `maxDurationSizeJSON` was slightly incorrect before - now it's the same as `maxDatetimeSizeJSON`. I've updated the comment and will do some further testing later on.

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/types.go
Comment (jiahuif) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/types.go:34]: Since it means "infinity" maybe it's better to make it `LONG_LONG_MAX`?

Comment (DangerOnTheRanger) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/types.go]: I like the idea of using the largest possible integer - I've changed the PR to use `math.MaxInt`, though I don't think it should be resolved just yet since I think using pointers may have merit too.

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:266]: The set of allowed types is not expected to change (changing it would be massively disruptive), so I'm okay with just returning noMaxLength here leaving a comment that it should be unreachable.

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/types.go
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/types.go]: I'm okay with either. Just remove any usages of -1 (there are a couple in the test code right now).

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (DangerOnTheRanger) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go]: Gotcha, fixed

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (cici37) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:111]: nit: `&&`

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (cici37) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:280]: Are you going to handle `byte`?

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (cici37) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:256]: How about `exclude required fields with a default specified since the request can omit them` raised in [comment](https://github.com/kubernetes/enhancements/pull/3144#discussion_r798043132)?

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (DangerOnTheRanger) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:280]: `byte` is handled by the `return 3` later on, since we don't do anything special for it compared to an otherwise-untyped string. I think @jpbetz left a review comment on simplifying this `if` statement since the `byte` check is essentially a no-op; I think that should address your comment as well

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (DangerOnTheRanger) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:256]: That's much more clear! I'll adjust the comment accordingly.

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (DangerOnTheRanger) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:255]: Good catch, I'll change it

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (DangerOnTheRanger) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:93]: Whoops - I'll fix that

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (DangerOnTheRanger) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:111]: Gotcha, I'll fix this as well

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go
Comment (cici37) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:280]: I didn't see `return 3` added here when `estimateMaxSizeJSON`

**Comment by DangerOnTheRanger** [comment]
/hold

#108419 is now the source of truth, this PR shouldn't be merged

**Pull request closed without merging**
